Store storage availability in database

RobinMcCorkell opened the pull request:
Storage status is saved in the database. Failed storages are rechecked every 10 minutes, while working storages are rechecked every request. Using the files_external app will recheck all external storages when the settings page is viewed, or whenever an external storage is saved.

The advantage of this is that failed external storages will not impede page loading, apart from every 10 minutes when a recheck is performed. To test, simply create several external storages will valid details other than the credentials (SFTP is nice and slow for this), then load the files page. On master, it will load slowly as timeouts occur. With this PR, it will load slowly once, then be fast for 10 minutes.

The counterpart to this is #13515, which eliminates the slowdown on the settings pages.

Fixes #13531

cc @PVince81 @icewind1991 @DeepDiver1975 

P.S. This is my first work involving database stuff. It's likely that I got something wrong, so review those bits in particular! The unit tests should help though :smile:

Commit:
Store storage availability in database

Storage status is saved in the database. Failed storages are rechecked every
10 minutes, while working storages are rechecked every request.

Using the files_external app will recheck all external storages when the
settings page is viewed, or whenever an external storage is saved.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/9b5d989d-4ec1-41d2-b250-36c89e69b394) was created.

LukasReschke commented on lib/private/files/cache/storage.php:
```diff
@@ -91,6 +99,36 @@ public static function getNumericStorageId($storageId) {
 	}
 
 	/**
+	 * @return bool
```
May also return null

Xenopathic replied:
Specifically cannot return null - see database schema

LukasReschke replied:
What is L108 then for?

Xenopathic replied:
Ah, sorry, it was late and I wasn't thinking properly :smile:

LukasReschke commented:
``` 
17:06:08 .....................................PHP Fatal error:  Call to undefined method Mock_Storage_77aba13b::test() in /var/jenkins/workspace/pull-request-analyser-ng-simple/label/SLAVE/lib/private/files/mount/mountpoint.php on line 207
17:06:11 PHP Stack trace:
17:06:11 PHP   1. {main}() /usr/local/bin/phpunit:0
17:06:11 PHP   2. PHPUnit_TextUI_Command::main() /usr/local/bin/phpunit:612
17:06:11 PHP   3. PHPUnit_TextUI_Command->run() phar:///usr/local/bin/phpunit/phpunit/TextUI/Command.php:138
17:06:11 PHP   4. PHPUnit_TextUI_TestRunner->doRun() phar:///usr/local/bin/phpunit/phpunit/TextUI/Command.php:186
17:06:11 PHP   5. PHPUnit_Framework_TestSuite->run() phar:///usr/local/bin/phpunit/phpunit/TextUI/TestRunner.php:423
17:06:11 PHP   6. PHPUnit_Framework_TestSuite->run() phar:///usr/local/bin/phpunit/phpunit/Framework/TestSuite.php:751
17:06:11 PHP   7. PHPUnit_Framework_TestCase->run() phar:///usr/local/bin/phpunit/phpunit/Framework/TestSuite.php:751
17:06:11 PHP   8. PHPUnit_Framework_TestResult->run() phar:///usr/local/bin/phpunit/phpunit/Framework/TestCase.php:722
17:06:11 PHP   9. PHP_Invoker->invoke() phar:///usr/local/bin/phpunit/phpunit/Framework/TestResult.php:641
17:06:11 PHP  10. call_user_func_array:{phar:///usr/local/bin/phpunit/php-invoker/Invoker.php:93}() phar:///usr/local/bin/phpunit/php-invoker/Invoker.php:93
17:06:11 PHP  11. PHPUnit_Framework_TestCase->runBare() phar:///usr/local/bin/phpunit/php-invoker/Invoker.php:93
17:06:11 PHP  12. PHPUnit_Framework_TestCase->runTest() phar:///usr/local/bin/phpunit/phpunit/Framework/TestCase.php:766
17:06:11 PHP  13. ReflectionMethod->invokeArgs() phar:///usr/local/bin/phpunit/phpunit/Framework/TestCase.php:881
17:06:11 PHP  14. Test\Files\Node\Root->testGet() phar:///usr/local/bin/phpunit/phpunit/Framework/TestCase.php:881
17:06:11 PHP  15. OC\Files\Node\Root->mount() /var/jenkins/workspace/pull-request-analyser-ng-simple/label/SLAVE/tests/lib/files/node/root.php:54
17:06:11 PHP  16. OC\Files\Mount\Manager->addMount() /var/jenkins/workspace/pull-request-analyser-ng-simple/label/SLAVE/lib/private/files/node/root.php:112
17:06:11 PHP  17. OC\Files\Mount\MountPoint->isAvailable() /var/jenkins/workspace/pull-request-analyser-ng-simple/label/SLAVE/lib/private/files/mount/manager.php:23
```

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/9b5d989d-4ec1-41d2-b250-36c89e69b394) completed: **18 updated code elements**

Xenopathic commented:
Hmm, I'm trying to fix these test failures, and it seems there is a cyclic dependency. This PR introduces the `isAvailable()` test before allowing a storage to be mounted, which checks if the storage can be `stat()`'d. However, during login, the filesystem is set up before the home storage is created (for a new user where it doesn't already exist), which fails because the home storage is unreachable at that stage.

There are a couple of solutions that I can see: either move the home storage creation so that it occurs before the filesystem is set up, or create the home storage while the storage is being mounted (so trigger that from `test()` or `isAvailable()`). This is going deep into things-may-break-horribly territory though, perhaps @icewind1991 can help?

PVince81 commented:
Nice work. I wonder if it's too early for this, as we wanted to revamp the DB anyway here: https://github.com/owncloud/core/issues/11261#issuecomment-61234239

But that might fit with the new scheme, too.
I'll review this later for 8.1.

DeepDiver1975 commented on lib/private/files/cache/storage.php:
```diff
@@ -91,6 +99,36 @@ public static function getNumericStorageId($storageId) {
 	/**
+	 * @return bool
+	 */
+	public function getAvailability() {
```
I'd merge this into one function where all information is returned. With the current design two queries will be fired to get availability and last checked

icewind1991 commented on lib/private/files/storage/wrapper/wrapper.php:
```diff
@@ -477,4 +477,31 @@ public function __call($method, $args) {
+	 * @return bool
+	 */
+	public function getAvailability() {
+		return $this->getStorageCache()->getAvailability();
```
```php
return $this->storage->getAvailability();
```
would be better here (and for the other 2 methods)

icewind1991 commented on lib/private/files/mount/manager.php:
```diff
@@ -20,7 +20,9 @@ class Manager {
 	 */
 	public function addMount(MountPoint $mount) {
-		$this->mounts[$mount->getMountPoint()] = $mount;
+		if ($mount->isAvailable()) {
```
:-1: 
This causes a call to `$storage->test()` (which can be very expensive) every time a storage is mounted whether it's used or not

Xenopathic replied:
Any suggestions for an alternative? Perhaps the only way to avoid this is wrapping every filesystem function with a check to `isAvailable()`?

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/2b4d06c2-833c-4b68-9bb5-308c9fe928dc) was created.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/96d371b3-b938-45f9-9543-d7d2e9951418) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/2b4d06c2-833c-4b68-9bb5-308c9fe928dc) completed: **2 new issues, 15 updated code elements**

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/96d371b3-b938-45f9-9543-d7d2e9951418) completed: **2 new issues, 15 updated code elements**

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/ad4b57d9-adcd-4987-bd23-8ef49c73fbbb) was created.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/1175e82c-71e5-45f6-9591-c5caafd7a948) was created.

Xenopathic commented:
Right, I think I've got it now. The availability checker is implemented as a storage wrapper, and uses the following algorithm for every storage call:

1. If marked as available in storage cache, continue with normal call.
2. If recheck TTL reached, perform storage test, update storage cache and go to 1. with new availability information
3. Otherwise, throw a `StorageNotAvailableException`

If the call to the storage itself throws a `StorageNotAvailableException`, the storage will be marked as not available in the storage cache, then the exception is rethrown up the call stack.

One issue noticed with this new code: `getDirectoryContent()` in a view will break if an exception is thrown at any point, such as when external storages visible in the root directory are probed. I added a try catch block to stop this, so that `StorageNotAvailableException`s are silently ignored.

For this PR to be useful, storages need to throw `StorageNotAvailableException` when they fail. At the moment, very few actually do, but for testing purposes the SFTP changes I made in #13390 work well. As mentioned earlier, SFTP is very good for noticing the timeouts, since an authentication failure typically has a long timeout.

@icewind1991 Tell me if I'm doing anything wrong :smile:

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/ad4b57d9-adcd-4987-bd23-8ef49c73fbbb) completed: **15 new issues, 46 updated code elements**

icewind1991 commented:
This approach seems fine by me

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/1175e82c-71e5-45f6-9591-c5caafd7a948) completed: **15 new issues, 46 updated code elements**

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/7e4d3273-3046-4c44-acb1-b03e613fafe5) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/7e4d3273-3046-4c44-acb1-b03e613fafe5) completed: **9 new issues, 54 updated code elements**

PVince81 commented:
Please rebase.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/bc3618a1-e9fd-49e1-af57-247268c8e309) was created.

Xenopathic commented:
Rebase conflict was the `version.php` update. No other code is changed.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/bc3618a1-e9fd-49e1-af57-247268c8e309) completed: **No issues found**

LukasReschke commented:
Rebase required.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/b644fa67-2560-438f-b100-7d55e6309ca6) was created.

Xenopathic commented:
Rebase conflict was license headers and version.php.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/b644fa67-2560-438f-b100-7d55e6309ca6) completed: **No issues found**

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/b644fa67-2560-438f-b100-7d55e6309ca6) completed: **No issues found**

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/eca992df-049d-4aef-a696-042181592579) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/eca992df-049d-4aef-a696-042181592579) completed: **11 new issues, 55 updated code elements**

Xenopathic commented:
@icewind1991 @LukasReschke @DeepDiver1975 @PVince81 Review please :smile:

LukasReschke commented on lib/private/files/storage/wrapper/availability.php:
```diff
@@ -0,0 +1,396 @@
+ * Throws a StorageNotAvailableException for storages with known failures
+ */
+class Availability extends Wrapper {
+	const RECHECK_TTL = 600; // 10 minutes
```
Maybe use something like `RECHECK_TTL_SEC` so that people don't have to look at the actual meaning of it :)

LukasReschke commented on lib/private/files/storage/wrapper/availability.php:
```diff
@@ -0,0 +1,396 @@
+		}
+	}
+
+	public function getDirectDownload($path) {
```
Inherit the PHPDocs so my PHPStorm with strict settings won't complain? :see_no_evil:

Xenopathic replied:
For you? Anything :laughing:

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/48be3077-eae0-4f94-a62d-217707ae822d) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/48be3077-eae0-4f94-a62d-217707ae822d) completed: **11 new issues, 54 updated code elements**

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/f5287564-2371-4d8f-8d4e-ec753acaf96e) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/f5287564-2371-4d8f-8d4e-ec753acaf96e) completed: **13 new issues, 54 updated code elements**

Xenopathic commented:
Hmm, that's the second time that's happened:

```
09:15:57 PHP  13. Test\Files\Filesystem->tearDown() phar:///usr/local/bin/phpunit/phpunit/Framework/TestCase.php:793
09:15:57 PHP  14. OC\Files\Filesystem::mount() /var/jenkins/workspace/pull-request-analyser-ng-simple/label/SLAVE/tests/lib/files/filesystem.php:56
09:15:57 PHP  15. OC\Files\Mount\MountPoint->__construct() /var/jenkins/workspace/pull-request-analyser-ng-simple/label/SLAVE/lib/private/files/filesystem.php:466
09:15:57 PHP  16. OC\Files\Storage\StorageFactory->wrap() /var/jenkins/workspace/pull-request-analyser-ng-simple/label/SLAVE/lib/private/files/mount/mountpoint.php:76
09:15:57 PHP  17. OC_Util::{closure:/var/jenkins/workspace/pull-request-analyser-ng-simple/label/SLAVE/lib/private/util.php:103-123}() /var/jenkins/workspace/pull-request-analyser-ng-simple/label/SLAVE/lib/private/files/storage/storagefactory.php:74
09:15:57 PHP  18. OC\Files\Storage\Wrapper\Wrapper->instanceOfStorage() /var/jenkins/workspace/pull-request-analyser-ng-simple/label/SLAVE/lib/private/util.php:110
09:15:57 PHP  19. OC\Files\Storage\Wrapper\Wrapper->instanceOfStorage() /var/jenkins/workspace/pull-request-analyser-ng-simple/label/SLAVE/lib/private/files/storage/wrapper/wrapper.php:455
09:15:57 PHP  20. OC\Files\Storage\Wrapper\Wrapper->instanceOfStorage() /var/jenkins/workspace/pull-request-analyser-ng-simple/label/SLAVE/lib/private/files/storage/wrapper/wrapper.php:455
09:15:57 PHP  21. OC\Files\Storage\Wrapper\Wrapper->instanceOfStorage() /var/jenkins/workspace/pull-request-analyser-ng-simple/label/SLAVE/lib/private/files/storage/wrapper/wrapper.php:455
09:15:57 PHP  22. OC\Files\Storage\Wrapper\Wrapper->instanceOfStorage() /var/jenkins/workspace/pull-request-analyser-ng-simple/label/SLAVE/lib/private/files/storage/wrapper/wrapper.php:455
... continue until function nesting level overload ...
```

Thing is, when this happened the last time, I scheduled a retest and it all worked fine.

@owncloud-bot retest this please

Xenopathic commented:
@owncloud-bot retest this please

Commit:
Fix mounting wrapped storages resulting in many-layered wrapping

This usually doesn't cause issues, but in unit tests sometimes a wrapped
storage is passed to Filesystem::mount() and gets rewrapped, hitting the
XDebug function nesting level limit when used.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/2c43c1b0-423e-4006-a19b-96777614a1b9) was created.

Xenopathic commented:
@icewind1991 Please check that latest commit to make sure I'm doing it right :smile:

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/2c43c1b0-423e-4006-a19b-96777614a1b9) completed: **7 new issues, 56 updated code elements**

MorrisJobke commented:
@Xenopathic This needs another rebase :( But it is moved to 8.2 ... maybe wait until 8.1 is out and then again try to work on this :)

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/1ac306ae-c629-48e2-a962-72a4aeba40ad) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/1ac306ae-c629-48e2-a962-72a4aeba40ad) completed: **10 new issues, 56 updated code elements**

icewind1991 commented on lib/private/util.php:
```diff
@@ -138,6 +138,11 @@ public static function setupFS($user = '') {
 
+			// install storage availability wrapper
+			\OC\Files\Filesystem::addStorageWrapper('oc_availability', function ($mountPoint, $storage) {
+				return new \OC\Files\Storage\Wrapper\Availability(['storage' => $storage]);
```
maybe dont apply the wrapper to local storages.

icewind1991 replied:
Maybe it's best to not register the wrapper here at all but have file_external add the wrapper to any external storage

Xenopathic replied:
@icewind1991 What about the internal DAV storage for S2S?

icewind1991 replied:
right, maybe then just have a check on `isLocal()`

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/84a4e344-abb8-4084-8525-1c52d8aa8522) was created.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/19c1a7ae-c223-4f8b-be38-111f7cb9145a) was created.

Xenopathic commented:
@owncloud-bot retest this please

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/19c1a7ae-c223-4f8b-be38-111f7cb9145a) completed: **No issues found**

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/19c1a7ae-c223-4f8b-be38-111f7cb9145a) completed: **No issues found**

icewind1991 commented:
:+1: nice work

MorrisJobke commented:
* I added a SMB backend in the admin panel
* green dot appears
* I shut down the SMB service
* I modified the last_accessed column to value-1000
* accessed the mounted folder to retrigger the check
* database still holds the available value `1`

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/cc0e72af-b966-457d-98ff-26b2839f0eab) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/cc0e72af-b966-457d-98ff-26b2839f0eab) completed: **37 new issues, 56 updated code elements**

Xenopathic commented:
@MorrisJobke This only works if the storage cooperates and throws StorageNotAvailableException. Try merging `sftp_exceptions` (#13390) with this branch and set up an SFTP storage. Then it will work as you expect. This PR just lays the groundwork for storage availability

MorrisJobke commented:
@Xenopathic rebase needed

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/2fd1b6ea-dd2c-4622-b1e2-a64c488cfcfa) was created.

Xenopathic commented:
@owncloud-bot retest this please

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/ee2d331b-029c-4bf1-b674-68b292cdbd66) was created.

Xenopathic commented:
@MorrisJobke Could you re-review this please?

Xenopathic commented:
Also cc @PVince81

PVince81 commented on lib/private/files/mount/mountpoint.php:
```diff
@@ -92,7 +93,11 @@ public function __construct($storage, $mountpoint, $arguments = null, $loader =
+			$this->storage = $storage;
+			// only wrap if not already wrapped
+			if (!($this->storage instanceof Wrapper)) {
+				$this->storage = $this->loader->wrap($this, $this->storage);
```
Is this required by this PR or is an additional fix ?

Xenopathic replied:
Without this fix, unit tests will fail, since our unit tests attempt to re-wrap storages sometimes causing function nesting limit failures. In the actual codebase it makes no difference, since we never pass a wrapped storage to be re-wrapped.

Xenopathic replied:
Actually, this is a poor fix. There may be usecases where instead of a Storage being passed in, a Wrapper is (aka pre-wrapped storage) but still needs to be wrapped with the other storage wrappers too.

PVince81 commented on lib/private/files/cache/storage.php:
```diff
@@ -130,6 +136,28 @@ public static function getNumericStorageId($storageId) {
+	 * @return array|null [ available, last_checked ]
+	 */
+	public function getAvailability() {
+		if ($row = self::getStorageById($this->storageId)) {
```
I guess it's not possible to retrieve + cache this together when reading the full storage entry ? (to avoid extra SQL calls)

Xenopathic replied:
The caching would be done in `getStorageById()`, since a couple of functions use that. But, prior to this PR, we never cached it (see `getNumericStorageId()` above, and the constructor). Do we want that in the scope of this PR?

PVince81 commented:
Didn't destroy the known universe :+1: (ran a few smashbox tests and encryption + ext storage)

PVince81 commented:
Regression, cannot create user: https://github.com/owncloud/core/issues/18129
